### ISSUE TITLE
Modbus Holding/Input Register Read Float - Incorrect response byte size

### ISSUE THREAD

**Issue by KyleKotowick** [zephyrproject-rtos/zephyr#68788]
**Describe the bug**
Using the Modbus server role, the input/holding register floating point read functions return an incorrect number of bytes. They return 64 bits (4 registers) instead of the 32 bits (2 registers) required for a 32-bit float.

A modbus client, when requesting a 32-bit float value, requests two 16-bit registers. Therefore, [`&ctx->rx_adu.data[2]`](https://github.com/zephyrproject-rtos/zephyr/blob/40b53d5d774705d27183aaa1e42f2f322250da10/subsys/modbus/modbus_server.c#L433) is `2`.

The Zephyr modbus library [then multiplies](https://github.com/zephyrproject-rtos/zephyr/blob/40b53d5d774705d27183aaa1e42f2f322250da10/subsys/modbus/modbus_server.c#L465) the number of registers (2) by the size of a float (32) and thinks it should return 64 bits. However, since the client only requested two 16-bit registers (32 bits), it gets an unexpected response size.

**To Reproduce**
1. Run a Zephyr modbus server, and enable `CONFIG_MODBUS_FP_EXTENSIONS`.
2. From a connected modbus client (in our case, an Automation Direct CM5-T7W), request a `float32` value from an address with value >= 5000.
3. See that the server tries to return 64 bits (4 registers) instead of the requested 2 registers (32 bits).

**Expected behavior**
When a 32-bit float is requested by a modbus client (two 16-bit registers), the modbus server should return 32 bits of data, not 64.

**Impact**
Showstopper. We cannot use the Zephyr modbus server with our Automation Direct CM5-T7W HMI, as we require the use of 32-bit floats.

**Environment (please complete the following information):**
OS: Windows 11
Toolchain: Zephyr 3.5.0

**Comment by jfischer-no** [comment]
Indeed, it looks suspicious, I am on it.

**Comment by KyleKotowick** [comment]
Not stale, problem still exists.

**Comment by KyleKotowick** [comment]
Not stale, problem still exists. @jfischer-no any luck?

**Comment by maksjapan** [comment]
Hi, I have the same problem (albeit with a different HMI).
From what I understand, the register quantity requested in the query header (reg_qty) is described in 2-byte length registers.
However, floats are in 4-byte registers, so the quantity of the floating point register should be counted as half of the register quantity requested in the query header.

I've added the below snipped in line 457 of the modbus_server.c: 
`reg_qty = (reg_qty + (sizeof(uint16_t) - 1))/sizeof(uint16_t);`

and it seems to fix the problem for Reading Input Registers (FC04).
I used the idea from [this discussion thread](https://stackoverflow.com/questions/2422712/rounding-integer-division-instead-of-truncating) in StackOverflow for the math.

I haven't tried it yet, but I presume for Reading Holding Registers (FC03), you'd add the same snipped in line 347 to fix it (assuming it has the same issue).
